storage/remote/remote_configurable.go

```diff
@@ -0,0 +1,116 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+// Storage collects multiple remote storage queues.
+type ConfigurableStorage struct {
+	mtx            sync.RWMutex
+	externalLabels model.LabelSet
+	conf           []config.RemoteWriteConfig
+
+	queues []*StorageQueueManager
+}
+
+// New returns a new remote Storage.
+func NewConfigurable() *ConfigurableStorage {
+	return &ConfigurableStorage{}
+}
+
+// ApplyConfig updates the status state as the new config requires.
+func (s *ConfigurableStorage) ApplyConfig(conf *config.Config) error {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	newQueues := []*StorageQueueManager{}
+	for _, c := range s.conf {
+		c, err := NewClient(c)
+		if err != nil {
+			return err
+		}
+		newQueues = append(newQueues, NewStorageQueueManager(c, nil))
+	}
+
+	for _, q := range s.queues {
+		q.Stop()
+	}
+	s.queues = newQueues
+	s.externalLabels = conf.GlobalConfig.ExternalLabels
+	s.conf = conf.RemoteWriteConfig
+	for _, q := range s.queues {
+		q.Run()
+	}
+	return nil
+}
+
+// Stop the background processing of the storage queues.
+func (s *ConfigurableStorage) Stop() {
+	for _, q := range s.queues {
+		q.Stop()
+	}
+}
+
+// Append implements storage.SampleAppender. Always returns nil.
+func (s *ConfigurableStorage) Append(smpl *model.Sample) error {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	var snew model.Sample
+	snew = *smpl
+	snew.Metric = smpl.Metric.Clone()
+
+	for ln, lv := range s.externalLabels {
+		if _, ok := smpl.Metric[ln]; !ok {
+			snew.Metric[ln] = lv
+		}
+	}
+
+	for _, q := range s.queues {
+		q.Append(&snew)
+	}
+	return nil
+}
+
+// NeedsThrottling implements storage.SampleAppender. It will always return
+// false as a remote storage drops samples on the floor if backlogging instead
+// of asking for throttling.
+func (s *ConfigurableStorage) NeedsThrottling() bool {
+	return false
+}
+
+// Describe implements prometheus.Collector.
+func (s *ConfigurableStorage) Describe(ch chan<- *prometheus.Desc) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	for _, q := range s.queues {
+		q.Describe(ch)
+	}
+}
+
+// Collect implements prometheus.Collector.
+func (s *ConfigurableStorage) Collect(ch chan<- prometheus.Metric) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	for _, q := range s.queues {
+		q.Collect(ch)
```

Comment:
The current codebase doesn't have this problem because every client type is only instantiated at most once.

The metrics should indeed have the same name (they already do in the current codebase), and the `type` label is already what differentiates the different queues. Just that we need to differentiate the multiple generic write queues now, which you're doing via the hostname prefix. Doesn't require changes to global/non-global states though...